src/Avalonia.Input/KeyboardNavigationHandler.cs

```diff
@@ -48,39 +49,24 @@ public void SetOwner(IInputRoot owner)
         {
             element = element ?? throw new ArgumentNullException(nameof(element));
 
-            var customHandler = element.GetSelfAndVisualAncestors()
-                .OfType<ICustomKeyboardNavigation>()
-                .FirstOrDefault();
+            // If there's a custom keyboard navigation handler as an ancestor, use that.
+            var custom = element.FindAncestorOfType<ICustomKeyboardNavigation>(true);
+            if (custom is object && HandlePreCustomNavigation(custom, element, direction, out var ce))
+                return ce;
 
-            if (customHandler != null)
+            var result = direction switch
             {
-                var (handled, next) = customHandler.GetNext(element, direction);
+                NavigationDirection.Next => TabNavigation.GetNextTab(element, false),
+                NavigationDirection.Previous => TabNavigation.GetPrevTab(element, null, false),
+                _ => throw new NotSupportedException(),
```

Comment:
It doesn't feels right, that we handle only two directions in this method. Up/Down/Left/Right should work as well at least. On the other hand I don't think we fully support arrow-keys navigation at all, so for now I believe it will be enough to update this method documentation to explicitly say about supported directions.

Reply:
Well to be nitpicky it does handle directional navigation if it finds an `ICustomKeyboardNavigation`, but what you say is true really. I actually have an implementation lying around somewhere that adds a directional search. I will try to find it and open a PR to add that.